proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,235 @@
+accounts on Solana Mainnet Beta. Given the relatively small number of vote
+rewards, it makes sense to keep vote rewards distribution mechanism unchanged. They can still be
+distributed efficiently at the first block of the epoch boundary. This reduce the
+impact on rewards for vote account and also simplify the overall changes, and let
+us focus on solving the primary bottleneck - Stake Rewards. Only Stake rewards are
+going to be distributed out over multiple blocks.
+
+In the new stake rewards distribution approach, we will separate the computation of
+rewards from the actual distribution of rewards at the epoch boundary by
+dividing the process into two distinct phases: (1) rewards calculation, and (2)
+rewards distribution.
+
+The "reward calculation" phase is defined as the blocks that fall under it are
+those where time is allocated to calculating the rewards, while the "rewards
+distribution" phase is defined as the blocks that fall under it are those where
+time is allocated to distributing the rewards.
+
+### Rewards Calculation
+
+The reward computation phase lasts for first `N` block heights since the
+beginning of the epoch. When reaching block height `N` after the start block
+of the `reward calculation phase`, the bank will mark it the end the
+`reward calculation phase`. A conservative value of 1,000 is chosen for `N` to
+accommodate the existing validators on Solana Mainnet Beta.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will block and wait until the reward computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase (end of block height `N`),
+`EpochRewardHistory` sysvar account and a set of `RewardReserve` sysvar
+accounts, which helps to track and verify the reward distribution progress, will be
+updated.
+
+### `EpochRewardHistory` Sysvar Account
+
+`EpochRewardHistory` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardHistory` is an array of `RewardHistoryEntry`,
+where each entry corresponds to the reward for one particular epoch. The
+maximum depth of the history are chosen to the same as the depth of stake history for consistency, which
+is 512.
+
+The layout of `EpochRewardHistory` sysvar is shown in the following pseudo code.
+
+```
+struct RewardHistoryEntry {
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
+   number_reward_reserve_accounts: u32,   // number of reward reserve accounts for reward distribution
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+The `EpochRewardHistory` is updated at the first block of the epoch boundary,
```

Comment:
To remove ambiguity, let's specify at the start or end of the first block. Whenever the current epoch reward calc/distribution occurs (I think at the start of the first block, before tx processing, right?)